stdlib/asm/crypto/fri.masm

```diff
@@ -0,0 +1,607 @@
+#!
+#! See https://github.com/0xPolygonMiden/miden-vm/issues/592#issuecomment-1341814867 for how τ is computed.
+#!
+#! Expected stack state
+#!
+#! [mem_ptr_q, mem_ptr_p, ...]
+#!
+#! Final stack state
+#!
+#! [τ1, τ0, mem_ptr_q, ...] | τ = (τ0, τ1)
+#!
+#! τ = rescue_prime::hash_elements(q... ++ p...)[..2]
+#!
+#! - `hash_elements` routine works over vector of elements of Fq | q = 2^64 -2^32 + 1
+#! - ... -> spread operator
+#! - ++ -> concatenate two vectors
+#! - q... -> vector of length 128 s.t. each element ∈ Fq
+#! - p... -> vector of length 16 s.t. each element ∈ Fq
+#! - Rescue Prime hash produces four Fq elements as its digest
+#! - [..2] -> picks first two Fq elements from digest
+proc.compute_tau
+    push.64
+    push.8
+    adv.ext2intt # interpolate polynomial p, using advice provider
+
+    drop
+    drop
+    dup
+
+    # Initialize Rescue state
+    push.144.0.0.0.0.0.0.0.0.0.0.0
+
+    # Consume `q` into Rescue state
+    repeat.16
+        mem_stream
+    end
+
+    movup.12
+    drop
+    movup.13
+    movdn.12
+
+    # Consume `p` into Rescue state
+    repeat.2
+        adv_pipe
+    end
+
+    movup.12
+    drop
+
+    dropw
+    swapw
+    dropw
+    drop
+    drop
+end
+
+#! Given ω, τ on stack top, this routine computes factor `ω/ (τ - ω)`
+#! of β, as written in formula https://github.com/0xPolygonMiden/miden-vm/pull/644#pullrequestreview-1265141094
+#! only when index j is even | j ∈ [0..64)
+#!
+#! Note, ω ∈ F_p | p = 2^64 - 2^32 + 1
+#!       τ ∈ Quadratic extension field over F_p
+#!
+#! Expected stack state looks like
+#!
+#! [ω, _, _, τ1, τ0, ...]
+#!
+#! Final stack state looks like
+#!
+#! [⍳1, ⍳0, _, _, τ1, τ0, ...] | ⍳ = ω/ (τ - ω)
+proc.compute_factor_of_beta_for_even_index
+    dup.4
+    dup.1
+    sub
+    dup.4
+    ext2inv
+    swap
+    dup.2
+    mul
+    movdn.2
+    mul
+end
+
+#! Given ω, τ on stack top, this routine computes factor `ω/ (τ - ω)`
+#! of β, as written in formula https://github.com/0xPolygonMiden/miden-vm/pull/644#pullrequestreview-1265141094
+#! only when index j is odd | j ∈ [0..64)
+#!
+#! Note, ω ∈ F_p | p = 2^64 - 2^32 + 1
+#!       τ ∈ Quadratic extension field over F_p
+#!
+#! Expected stack state looks like
+#!
+#! [ω, _, _, _, _, τ1, τ0, ...]
+#!
+#! Final stack state looks like
+#!
+#! [⍳1, ⍳0, _, _, _, _, τ1, τ0, ...] | ⍳ = ω/ (τ - ω)
+proc.compute_factor_of_beta_for_odd_index
+    dup.6
+    dup.1
+    sub
+    dup.6
+    ext2inv
+    swap
+    dup.2
+    mul
+    movdn.2
+    mul
+end
+
+#! Given ⍳, accumulator ν on stack top, this routine accumulates ⍳ into ν only when j is even | j ∈ [0..64)
+#! following formula described in https://github.com/0xPolygonMiden/miden-vm/pull/644#pullrequestreview-1265141094
+#!
+#! Expected stack top
+#!
+#! [ω, ν1, ν0, τ1, τ0, mem_ptr_q - 1, ...]
+#!
+#! Final stack state
+#!
+#! [a3, a2, ν1', ν0', τ1, τ0, mem_ptr_q, ...]
+proc.accumulate_for_even_index
+    exec.compute_factor_of_beta_for_even_index
```

Comment:
I would get rid of `compute_factor_of_beta_for_even_index` and would just inline the code here. If we add some comments about start and end stack state, I think this would actually be more readable than having a separate function.

Reply:
Makes sense, fixed now.